lib/semantics/check-do.cc

```diff
@@ -52,6 +52,20 @@ static void SayWithDo(SemanticsContext &context, parser::CharBlock stmtLocation,
   context.Say(stmtLocation, message).Attach(doLocation, GetEnclosingDoMsg());
 }
 
+static void SayWithDeclaration(SemanticsContext &context,
+    const parser::CharBlock &location, const parser::MessageFixedText &&message,
+    const parser::CharBlock &name, const Symbol &symbol) {
+  if (const auto *details{symbol.detailsIf<UseDetails>()}) {
+    context.Say(location, message, name)
+        .Attach(details->location(),
+            "It is use-associated with '%s' in module '%s'"_err_en_US,
+            details->symbol().name(), details->module().name());
+  } else {
+    context.Say(location, message, name)
+        .Attach(symbol.name(), "Declaration of '%s'"_en_US, symbol.name());
+  }
+}
+
```

Comment:
This function seems generally useful, so I think it should be in `SemanticsContext` along with the `Say` it is calling. I suggest an interface like this:
```
template<typename... A>
void SayWithDecl(const Symbol &symbol, parser::CharBlock at,
    parser::MessageFixedText &&msg, A &&... args);
```

We use "Declaration" and "Decl" in names throughout the compiler, but "Decl" appears to be more common and has the advantage of being shorter.

Reply:
Thanks, Tim.  I'll do that.